### PR TITLE
fix: unhandled exception resulting from missing config sections 

### LINKED ISSUE #2179: Removing any entry in the envrionment configs results in a crash
# Bug

If the `redis` or any other configuration (even if redis is not enabled) is removed from the config the server crashes on startup. 

With an obscure error message 

```
Unhandled exception:
type 'Null' is not a subtype of type 'Map<dynamic, dynamic>'
```

 ## Expected behavior

Should not crash and should probably not be required unless this feature is enabled.

If the feature is enabled but the config is missing, give a reasonable error message.

### Discussed in https://github.com/serverpod/serverpod/discussions/2089

<div type='discussions-op-text'>

<sup>Originally posted by **AslamThachapalli** March 31, 2024</sup>
The purpose of docker is to make a pg db up and running right. 
Let's say I don't want to use docker, but a hosted db from, say neon.tech or aiven, for the dev environment. Isn't it straight forward and possible?

I added the following config in development.yaml:
```yaml
database:
  host: ep-hottest-noice-a16o6wqg.ap-southeast-1.aws.neon.tech
  port: 5432
  name: ptest
  user: aslam.develop912%50gmail.comd

# This is the setup for Redis.
# redis:
#   enabled: false
#   host: localhost
#   port: 8091
```
and edited the passwords.yaml
```yaml
development:
  database: 'thisIsADummyPass'
  # redis: '5mXNVdStJLrTvVQTpvA8Sv_NaTuOBT_e'
```
But when I do `dart bin/main.dart --apply-migrations`, I am getting the following error:
```
Unhandled exception:
type 'Null' is not a subtype of type 'Map<dynamic, dynamic>'
#0      new ServerpodConfig (package:serverpod_shared/src/config.dart:65:9)
#1      new Serverpod (package:serverpod/src/server/serverpod.dart:271:14)
#2      run (package:mypod_server/server.dart:14:15)
#3      main (file:///Users/aslam/Programming/Serverpod/mypod/mypod_server/bin/main.dart:6:3)
#4      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
#5      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```
---
But when I tried with the docker setup everything was working fine though. Is there anything I am missing?</div>

### PR DESCRIPTION
**Pull request by bfemmer**
This PR addresses the unhandled exception that occurs when a YAML config file is missing configuration sections (such as Redis) when Serverpod starts up. The root cause of the bug is the null assignment to a Map variable as a result of the null output of the yaml _doc_ command as a result of the parsing of an incomplete config file, highlighted in the screenshot below:

image

The rationale for the proposed fix is to implement null operator (??) on the doc method and return an empty map to the Map variable if null (indicating a missing config) and then subsequently evaluate the map contents ... if empty, assign null (for all servers except API server ... see below); otherwise, assign the results of the applicable __fromJson()_ method.

For the API server, the field is defined as not nullable in the ServerpodConfig class, so if the config is missing, a default ServerConfig instance is assigned.

Of note is the webServer field when a configuration is missing, which throws an unhandled exception with a very good and detailed description ...

image

This is the same exception that is thrown when the deployment.yaml configuration file is outright missing ...

image

The webServer is the only server that throws this exception and this appears to be operating as designed. If the expected behavior should be different, recommend a separate issue to address implementing new behavior.

No additional tests were added for this update, all asserts were left in place.

Fixes #2179 

### Addition by @SandPod 

Added validation for all configuration fields with helpful messages.
Fixes: #2294

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

No breaking changes were detected or observed during testing of the above patch.

### CODE REVIEW

**Comment by bfemmer** [comment]
recheck

**Comment by bfemmer** [comment]
I have read the CLA Document and I hereby sign the CLA

File: packages/serverpod_shared/lib/src/config.dart [hunk 1]
```diff
@@ -74,14 +74,30 @@ class ServerpodConfig {
 
     var doc = loadYaml(data);
 
+    /// Get api server setup. This field cannot be null, so if the
+    /// configuration is missing, create a default configuration.
     assert(doc['apiServer'] is Map, 'apiServer is missing in confing');
-    var apiServer = ServerConfig._fromJson(doc['apiServer']);
-
+    Map apiSetup = doc['apiServer'] ?? {};
+    var apiServer = apiSetup.isNotEmpty
```
Comment (Isakdl) [inline packages/serverpod_shared/lib/src/config.dart:81]: Suggestion: Make an explicit nullcheck here and throw if the `apiServer` key is missing, as it is a required field. Remove the assert.

File: packages/serverpod_shared/lib/src/config.dart [hunk 2]
```diff
@@ -74,14 +74,30 @@ class ServerpodConfig {
...
+        ? ServerConfig._fromJson(apiSetup)
+        : ServerConfig(
+            port: 8080,
+            publicHost: 'localhost',
+            publicPort: 8080,
+            publicScheme: 'http',
+          );
```
Comment (Isakdl) [inline packages/serverpod_shared/lib/src/config.dart:88]: Sugggestion: Remove the default. If you specify a file you have to also give a config for the apiServer this makes it explicit and clear for the developer.

File: packages/serverpod_shared/lib/src/config.dart [hunk 3]
```diff
@@ -74,14 +74,30 @@ class ServerpodConfig {
...
+
+    /// Get insights server setup
     assert(doc['insightsServer'] is Map, 'insightsServer is missing in config');
-    var insightsServer = ServerConfig._fromJson(doc['insightsServer']);
+    Map insightsSetup = doc['insightsServer'] ?? {};
+    var insightsServer =
+        insightsSetup.isNotEmpty ? ServerConfig._fromJson(insightsSetup) : null;
```
Comment (Isakdl) [inline packages/serverpod_shared/lib/src/config.dart:94]: Suggestion: Change this to a null check instead e.g.

```dart
Map insightsSetup = doc['insightsServer'];
    var insightsServer =
        insightsSetup != null ? ServerConfig._fromJson(insightsSetup) : null;
```

This way we allow the `insightsServer` keyword to be optional BUT if you specify it you have to also give a valid config! 

This comment applies to all the other configurations in this file too.

**Review by Isakdl — CHANGES_REQUESTED** [review]
Thank you for the PR! :) 

Very nice addition and improvement. I do have some minor requested changes before we merge it!

**Comment by SandPod** [comment]
@bfemmer I'll fix the last comments here so we can get this in.

File: packages/serverpod_shared/test/config_test.dart [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    expect(config.apiServer.port, 8080);
+    expect(config.apiServer.publicHost, 'localhost');
+    expect(config.apiServer.publicPort, 8080);
+    expect(config.apiServer.publicScheme, 'http');
+    expect(config.insightsServer, isNull);
+    expect(config.webServer, isNull);
+    expect(config.maxRequestSize, 524288);
+    expect(config.serviceSecret, 'password');
+    expect(config.database, isNull);
+    expect(config.redis, isNull);
```
Comment (Isakdl) [inline packages/serverpod_shared/test/config_test.dart:104]: Suggestion: Refactor this test into a group and specific tests for each slice, this makes it easier to add / remove things in the future if the structure changes and/or grows and shrinks. Ideally we shouldn't have to edit expects but rather delete / remove the entire test, leads to less risk of changing the wrong thing.

File: packages/serverpod_shared/test/config_test.dart [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+  port: 8080
+  publicHost: localhost
+  publicPort: 8080
+  publicScheme: http
+''';
+
+    var config = ServerpodConfig.loadFromYaml(
+      'myRunMode',
+      'myServerId',
+      {'serviceSecret': 'password'},
```
Comment (Isakdl) [inline packages/serverpod_shared/test/config_test.dart:89]: Suggestion: Put a longer password here, even if it doesn't matter for this test we require a longer password later on in the chain.

File: packages/serverpod_shared/test/config_test.dart [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+        (e) => e.toString(),
+        'message',
+        equals(
+            'Exception: apiServer configuration has invalid type for port. Expected int, got String.'),
+      )),
+    );
+  });
+
+  test(
+      'Given Serverpod config with api server configuration when loading from Yaml then config is loaded correctly.',
```
Comment (Isakdl) [inline packages/serverpod_shared/test/config_test.dart]: Suggestion: `is loaded correctly` is ambiguous be specific for each test. This comment should be read together with the one below on the point of the expects.

File: packages/serverpod_shared/test/config_test.dart [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+      () {
+    var serverpodConfig = '''
+apiServer:
+  port: 8080
+  publicHost: localhost
+  publicPort: 8080
+  publicScheme: http
+redis:
+  host: localhost
+  port: 6379
```
Comment (Isakdl) [inline packages/serverpod_shared/test/config_test.dart:318]: Question: Redis also takes the key/value pair `enabled: bool` to toggle the config. Do we need a test for that here?

File: packages/serverpod_shared/test/config_test.dart (hunk 4 above)
Comment (Isakdl) [reply packages/serverpod_shared/test/config_test.dart:104]: Apply this logic on all tests.

**Review by Isakdl — CHANGES_REQUESTED** [review]
Looking good! I have a few comments on the test structure but otherwise ready to go.

File: packages/serverpod_shared/test/config_test.dart (hunk 7 above)
Comment (SandPod) [reply packages/serverpod_shared/test/config_test.dart:318]: The value doesn't toggle if the configuration is read or not. It is only a field on the configuration object with the default value false. Will treat it as other things with default values.

**Review by Isakdl — APPROVED** [review]
LGTM

**Pull request merged**